pkg/utils/tsoutil/tso_dispatcher.go

```diff
@@ -45,119 +43,159 @@ type TSODispatcher struct {
+		return val.(error)
+	}
+	return nil
+}
+
+// DispatchRequest is the entry point for dispatching/forwarding a tso request to the destination host
+func (s *TSODispatcher) DispatchRequest(req Request, tsoProtoFactory ProtoFactory) {
 	val, loaded := s.dispatchChs.LoadOrStore(req.getForwardedHost(), make(chan Request, maxMergeRequests))
 	reqCh := val.(chan Request)
 	if !loaded {
-		tsDeadlineCh := make(chan deadline, 1)
-		go s.dispatch(ctx, tsoProtoFactory, req.getForwardedHost(), req.getClientConn(), reqCh, tsDeadlineCh, doneCh, errCh, tsoPrimaryWatchers...)
-		go watchTSDeadline(ctx, tsDeadlineCh)
+		go s.startDispatchLoop(req.getForwardedHost(), req.getClientConn(), reqCh, tsoProtoFactory)
 	}
 	reqCh <- req
 }
 
-func (s *TSODispatcher) dispatch(
-	ctx context.Context,
-	tsoProtoFactory ProtoFactory,
-	forwardedHost string,
-	clientConn *grpc.ClientConn,
-	tsoRequestCh <-chan Request,
-	tsDeadlineCh chan<- deadline,
-	doneCh <-chan struct{},
-	errCh chan<- error,
-	tsoPrimaryWatchers ...*etcdutil.LoopWatcher) {
+// startDispatchLoop starts the dispatch loop for the forwarded host
+func (s *TSODispatcher) cleanup(
+	forwardedHost string, finalForwardErr error,
+	unprocessedRequests []Request, unprocessedReqCount int,
+) {
+	pendingReqCount := unprocessedReqCount
+	pendingRequests := unprocessedRequests[:unprocessedReqCount]
+	val, loaded := s.dispatchChs.LoadAndDelete(forwardedHost)
+	if loaded {
+		reqCh := val.(chan Request)
+		waitingReqCount := len(reqCh)
+		for i := 0; i < waitingReqCount; i++ {
+			req := <-reqCh
+			pendingRequests = append(pendingRequests, req)
+			pendingReqCount++
+		}
+	}
+	if finalForwardErr != nil {
+		for i := 0; i < pendingReqCount; i++ {
+			if pendingRequests[i] != nil {
+				pendingRequests[i].sendErrorResponseAsync(finalForwardErr)
+			}
+		}
+	} else if pendingReqCount > 0 {
+		log.Warn("the dispatch loop exited with pending requests unprocessed",
+			zap.String("forwarded-host", forwardedHost),
+			zap.Int("pending-requests-count", pendingReqCount))
+	}
+}
+
+func (s *TSODispatcher) startDispatchLoop(
+	forwardedHost string, clientConn *grpc.ClientConn,
+	tsoRequestCh <-chan Request, tsoProtoFactory ProtoFactory,
+) {
 	defer logutil.LogPanic()
-	dispatcherCtx, ctxCancel := context.WithCancel(ctx)
-	defer ctxCancel()
-	defer s.dispatchChs.Delete(forwardedHost)
+	ctx, cancel := context.WithCancel(s.ctx)
+	defer cancel()
+
+	// forwardErr indicates the failure in the forwarding stream which causes the dispatch loop to exit.
+	var (
+		forwardErr    error
+		forwardStream stream
+	)
+	pendingRequests := make([]Request, maxMergeRequests+1)
+	pendingTSOReqCount := 0
+
+	log.Info("start the dispatch loop", zap.String("forwarded-host", forwardedHost))
+	defer func() {
+		log.Info("exiting from the dispatch loop. cleaning up the pending requests",
+			zap.String("forwarded-host", forwardedHost))
+		if forwardStream != nil {
+			forwardStream.closeSend()
```

Comment:
Does it mean that a new stream is created for each request?

Reply:
It won't, because we only start one dispatch loop (this function) for all requests with the same forwarded host until there is any forwarding related error which causes the loop to exit, and we do this check and forwardStream.closeSend() in defer func() when exiting from this loop.